Property Reference Underlying Instance Type

333fred opened the pull request:
Fixes https://github.com/dotnet/roslyn/issues/21769, by making the type of the implicit IInstanceReferenceExpression match the type of the object being constructed, rather than the type defining the property. In order to do this in C#, I had to add a symbol to the bound node. I also added support for VB, as it currently just has NoneOperation for this node.

333fred commented:
@dotnet-bot retest windows_debug_unit32_prtest please

jinujoseph commented:
test windows_debug_unit32_prtest

333fred commented:
Ping @dotnet/analyzer-ioperation and @dotnet/roslyn-compiler for some reviews please.

cston commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -535,7 +535,7 @@ private IMemberReferenceExpression CreateBoundObjectInitializerMemberOperation(B
                 semanticModel: _semanticModel,
                 syntax: boundObjectInitializerMember.Syntax,
-                type: boundObjectInitializerMember.MemberSymbol.ContainingType,
+                type: boundObjectInitializerMember.InstanceSymbol,
```
Would it be sufficient to walk up to the containing `BoundObjectInitializerExpression` and use the `Type` of that node instead?

333fred replied:
How though? We don't have a parent operation here to use, and doing that via syntax feels like it would be very fragile to me.

cston replied:
I see. In that case, holding on to the receiver type makes sense.

cston commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:
```diff
@@ -1322,6 +1322,7 @@
     <Field Name="Expanded" Type="bool"/>
     <Field Name="ArgsToParamsOpt" Type="ImmutableArray&lt;int&gt;" Null="allow"/>
     <Field Name="ResultKind" PropertyOverrides="true" Type="LookupResultKind"/>
+    <Field Name="InstanceSymbol" Type="TypeSymbol" Null="disallow"/>
```
The term `Instance` is not used elsewhere in `BoundNodes`. Perhaps rename to `ReceiverType`.

333fred replied:
Updated.

333fred commented:
@cston addressed feedback.

cston commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:
```diff
@@ -1322,6 +1322,9 @@
     <Field Name="ResultKind" PropertyOverrides="true" Type="LookupResultKind"/>
+
+    <!-- Used by IOperation to reconstruct the receiver for this expression. -->
+    <Field Name="ReceiverSymbol" Type="TypeSymbol" Null="disallow"/>
```
Consider `ReceiverType` rather than `ReceiverSymbol`. The latter sounds like the actual receiver (for instance, the `LocalSymbol` being created).

AlekseyTs replied:
> Consider ReceiverType rather than ReceiverSymbol. The latter sounds like the actual receiver (for instance, the LocalSymbol being created).

Or TypeBeingInitialized?

AlekseyTs commented:
@333fred It looks like there are legitimate test failures

AlekseyTs commented:
>                     return boundMember;

Do we have coverage for this code path?

---
Refers to: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:3994 in 9d32572. [](commit_id = 9d3257299252cbc345d2ff1e718a1312249dce4e, deletion_comment = False)

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/MethodToClassRewriter.cs:
```diff
@@ -633,7 +633,7 @@ public override BoundNode VisitObjectInitializerMember(BoundObjectInitializerMem
             }
 
-            return node.Update(member, arguments, node.ArgumentNamesOpt, node.ArgumentRefKindsOpt, node.Expanded, node.ArgsToParamsOpt, node.ResultKind, node.BinderOpt, type);
+            return node.Update(member, arguments, node.ArgumentNamesOpt, node.ArgumentRefKindsOpt, node.Expanded, node.ArgsToParamsOpt, node.ResultKind, node.ReceiverSymbol, node.BinderOpt, type);
```
>node.ReceiverSymbol [](start = 153, length = 19)

We probably should visit this type - ```this.VisitType(node.ReceiverSymbol)```

333fred replied:
Fixed.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IPropertyReferenceExpression.cs:
```diff
@@ -10,7 +10,7 @@ public partial class IOperationTests : SemanticModelTestBase
         [CompilerTrait(CompilerFeature.IOperation)]
         [Fact]
-        public void IPropertyReferenceExpression_PropertyReferenceInDerivedTypeUsesBaseTypeAsInstanceType()
+        public void IPropertyReferenceExpression_PropertyReferenceInDerivedTypeUsesDerivedTypeAsInstanceType()
```
>IPropertyReferenceExpression_PropertyReferenceInDerivedTypeUsesDerivedTypeAsInstanceType [](start = 20, length = 88)

Consider adding a WorkItem for the issue being fixed.

333fred replied:
Done.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb:
```diff
@@ -716,6 +720,22 @@ Namespace Microsoft.CodeAnalysis.Semantics
+            Return New InstanceReferenceExpression(_semanticModel, syntax, type, constantValue, isImplicit)
+        End Function
+
+        Private Function CreateBoundWithRValueExpressionPlaceholder(boundWithRValueExpressionPlaceholder As BoundWithRValueExpressionPlaceholder) As IInstanceReferenceExpression
```
>CreateBoundWithRValueExpressionPlaceholder [](start = 25, length = 42)

It look like there is no test coverage for this new code, it appears BoundWithRValueExpressionPlaceholder is only used with ```With``` statement.

333fred replied:
There are a few: https://github.com/dotnet/roslyn/pull/22076/files/9d3257299252cbc345d2ff1e718a1312249dce4e#diff-f63cce822741d92606b3910d4c4d5e14L56 as an example. I also added https://github.com/dotnet/roslyn/pull/22076/files/9d3257299252cbc345d2ff1e718a1312249dce4e#diff-a9ede0042063c66b3a49be3a727f7260R42 for this scenario.

AlekseyTs replied:
> There are a few

Please confirm at least one of the the ```With``` statement scenarios is actually using BoundWithRValueExpressionPlaceholder node, it is hard to tell whether this is the case.

333fred replied:
I confirmed it debugging though the code manually.

AlekseyTs commented:
Done with review pass (iteration 6).

333fred commented:
@AlekseyTs addressed feedback. Added https://github.com/dotnet/roslyn/pull/22076/files#diff-bd1c2b37f6758b5aa11e137960887ac7R329 for the case you brought up.

333fred commented:
Because of the cloning we're now doing for arrays, I believe this is going to need @heejaechang's GetOperation rework to pass.

333fred commented:
Ping @dotnet/analyzer-ioperation for a review please.

333fred commented:
Also, @AlekseyTs, I didn't add CreateBoundWithRValueExpressionPlaceholder until I noticed that some of the VB tests still had NoneOperation nodes in them when I was expecting them to be fixed. I've verified that the code is being hit.
